src/community/wps-longitudinal-profile/src/main/java/org/geoserver/wps/longitudinal/LongitudinalProfileProcess.java

```diff
@@ -174,27 +170,79 @@ public LongitudinalProfileProcessResult execute(
         Geometry previousPoint = null;
 
         CoordinateReferenceSystem coverageCrs = coverageInfo.getCRS();
-        List<Position2D> positions2D =
-                Arrays.stream(denseLine.getCoordinates())
-                        .map(coordinate -> new Position2D(coverageCrs, coordinate.x, coordinate.y))
-                        .collect(Collectors.toList());
 
-        for (Position2D position2D : positions2D) {
-            LOGGER.fine("processing position:" + position2D);
+        // Create a stack with all geometry vertices
+        Coordinate[] coords = denseLine.getCoordinates();
+        Stack<ProfileVertice> vertices = new Stack<ProfileVertice>();
+        for (int i = 0; i < coords.length; i++) {
+            vertices.add(
+                    new ProfileVertice(
+                            i, new Position2D(coverageCrs, coords[i].x, coords[i].y), null));
+        }
+
+        // Divide vertices stack in chunks to process altitude reading with threads depending of
+        // available cores
+        int cores = Runtime.getRuntime().availableProcessors();
+        int total = vertices.size();
+        int chunkSize = total / cores;
+        int overflow = total % cores;
+
+        Stack<Vector<ProfileVertice>> chunks = new Stack<Vector<ProfileVertice>>();
+        for (int i = 0; i < cores; i++) {
+            Vector<ProfileVertice> chunk = new Vector<ProfileVertice>();
+            for (int j = 0; j < chunkSize; j++) {
+                chunk.add(vertices.pop());
+            }
+            if (overflow > 0) {
+                chunk.add(vertices.pop());
+                overflow = overflow - 1;
+            }
+            chunks.add(chunk);
+        }
+
+        // Create threads executor
+        ExecutorService executor = Executors.newFixedThreadPool(cores);
```

Comment:
Creating a separate pool for each process invocation is source for trouble, if a few WPS request pile up you'll get ``N*cores`` threads running. Also, each thread queries the data sources, which would easily lead to starvation of the connection pools. Long story short, as written this change is surely going to benefit a server with light load, but would crush it once the load goes up.

A better version would be to have a single thread pool for the whole process (held as a field), so that light load gets more cores, but higher loads cannot diverge. 

But generally speaking, it's best not to have extra threads. For example, if the time is mostly spent querying the data sources, better to accumulate a batch of coordinates to adjust (not too many, to avoid OOMs) and do a single query grabbing all adjustment features instead.

Reply:
> A better version would be to have a single thread pool for the whole process (held as a field), so that light load gets more cores, but higher loads cannot diverge.

I'm trying to move the thread pool to class level as you suggest.

Have you some suggestions for manage the shutdown ?
Should I just put it in finalise() ? 
Or use application event like there : https://github.com/geoserver/geoserver/blob/45de065ff2219ac754fb567949169f57c629b74f/src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/WPSExecutionManager.java#L330-L339 ?

> For example, if the time is mostly spent querying the data sources, better to accumulate a batch of coordinates to adjust (not too many, to avoid OOMs) and do a single query grabbing all adjustment features instead.

Not sure to understand this part. The code for the adjustment part is the same as already existing one. I just put it in thread to perform the adjustment directly after altitude reading. But the amount of vertice to read and adjust is exactly the same as current code. It depends on the step and geometry params. Maybe it can be improved, but not related to this PR.

Reply:
There are two choices here:
* Keep the algorithm as is, use multiple threads. This is the brute-force way of addressing the issue.
* Improve the existing algorithm so that it does not do as many queries, thus, less back and forth between the process and the adjustment data source. This should lead to improved performance without having to call for extra threads (work smarter, not harder)

The current code adjusts one point at a time, and as a result, makes a query to the adjustment source per point. Each query has a communication latency and an execution time. The total latency here is num points * single query latency, can pile up and become dominant.
If you collect a bunch of points (e.g., 100) and query the adjustment source for an intersection over all of them, you pay 1 latency only for that query: the overall latency is reduced by a factor of 100.

The downside of this approach is that it requires more complex code, the upside is that it will speed up execution without requiring multiple threads (overall, it will scale up better).

One final note: while I'd guess the second approach will be better overall (better performance, better scalability) , I cannot be sure of it. A profiling session would be needed, to check where the time is actually spent (if you see it's spent communicating with the database, then you have proof that latency is the source of slowness).

Reply:
Ok, but the aim of my PR is to improve reading of the altitude from coverage by using multiple threads.
In the original WPS process I develop (on which this one is based) I manage to improve the process duration with this method.

The aim of this PR is to introduce threading to process vertices faster.

The adjustment function is introduced with this WPS process, and I agree it could be improved, but is not related to this PR.

I think there are multiple vector of improvment that are not exclusive : 

- Use multiple threading to read raw altitude (what this PR intend to do), becaus we must read this value for all vertices
- Improve the adjustment process by trying to perform it on group of vertices (to do in another dedicated PR)